include/seqan3/argument_parser/argument_parser.hpp

```diff
@@ -416,6 +416,24 @@ class argument_parser
         return *sub_parser;
     }
 
+    /*!\brief Checks whether the option distinguished by `id` was set on the command line by the user.
+     * \param[in] id The option identifier to search for (must not contain dashes!).
+     * \returns `true` if option identifier `id` was set on the command line by the user.
+     */
+    bool is_option_set(char const id)
+    {
+        auto end = std::find(cmd_arguments.begin(), cmd_arguments.end(), "--");
+        return detail::format_parse::find_option_id(cmd_arguments.begin(), end, id) != end;
```

Comment:
I'm having trouble understanding what you are doing.

1. Why search for `--` when you are looking at short ids?
2. Wouldn't you want to use `std::find_end` since you surely don't want `end` to point to the first occurrence of `--`?
3. Why don't you just pass begin and end of cmd_arguments?

Reply:
I just noticed that you answered yesterday under my month old comment, maybe you can add the reasoning for the `--` as a comment, because it is really confusing.
Also, where does the `--` come from?

Reply:
> No this is something different.
I actually do want to search for the first occurrence of "--". Note: Not as a prefix (or other -fix) of another word but ONLY "--" because it signals the end of options and marks the beginning of arguments for positional options. It is special in its meaning. And since I am searching for option names I must only search before the arguments for positional options.
Was that explained in an understandable way? Otherwise I'll try again.

`--` is a POSIX convention to singal that there are no more option names after this. That way the user can use dash in arguments without confusing the argument parser. This done like this in most argument parsers.

See [POSIX conventions](http://www.iitk.ac.in/esc101/05Aug/tutorial/essential/attributes/_posix.html)

Reply:
Do you add the `--` somewhere or is it magically there?

Reply:
It is user input on the command line that may be there

Reply:
Phew maybe in the tutorial. I thought this is common knowledge somehow. 
I will make a follow up issue to make sure the parsing conventions are mentioned somewhere. It oculd just be I said somewhere "we follow official POSIX conventions".

Reply:
Here is the issue https://github.com/seqan/product_backlog/issues/201

Reply:
And while looking at it. You should make `--` a static constexpr variable somewhere in the class's scope such that it also gets a proper name: `end_of_options_marker` or something like that.

Reply:
some more thoughts about that. If I understand this correctly, then we have to call parse before we can ask whether an option was set by the user. So should it not throw in the case parse wasn't called before? I think we have a variable that caches this?

Reply:
The other thing I'd like to reconsider from the implementation is the duplication of this code. 
First, I observed that an extra copy is made to store the parsed arguments in the argument parser. 
Second, the arguments are only necessary if format_parse was called so also this option can only be available if format_parse was called.

So the code should be refactored that I can ask for the set option in the format_parse class. Calling it for any other parsing option can then throw, right? So since all formats inherit from a base class this can be achieved easily. Then only the format_parse class overloads this and inside of the argument_parser you call std::visit on the selected format.

Reply:
> some more thoughts about that. If I understand this correctly, then we have to call parse before we can ask whether an option was set by the user. So should it not throw in the case parse wasn't called before? I think we have a variable that caches this?

No, we do not **have to** call parse before checking this. It is independent. (see test case)
This is the case because we have the list `argv` already on construction will all the information we need.
We can of course enforce this restriction anyway if we think it is confusing for the user or more intuitive :shrug:.

> The other thing I'd like to reconsider from the implementation is the duplication of this code.

Which duplication?

> First, I observed that an extra copy is made to store the parsed arguments in the argument parser.

Yes. We now have to copies of `argv`. One in the `argument_parser` and one in `format_parse`.
But this is needed either way, even if I move this list into `format_parse` because format_parse is stripping it's copy of `argv` of arguments while parsing the information s.t. it can afterwards detect "too many" or "unknown option identifiers". So in order to enable `is_option_set` I need to store the original `argv` in some place anyway.

> Second, the arguments are only necessary if format_parse was called so also this option can only be available if format_parse was called.

Technically no, see above.

> So the code should be refactored that I can ask for the set option in the format_parse class. Calling it for any other parsing option can then throw, right? So since all formats inherit from a base class this can be achieved easily. Then only the format_parse class overloads this and inside of the argument_parser you call std::visit on the selected format.

IMHO I think that `is_option_set` is not strongly attached to `format_parse`. It's a feature of the argument parser that doesn't involve actual parsing of argument values but only checking if an id can be found. Is this such a strong design decision, whether we call a static function of format_parse or use std::visit?

Reply:
> No, we do not have to call parse before checking this. It is independent. (see test case)
> IMHO I think that is_option_set is not strongly attached to format_parse. It's a feature of the argument parser that doesn't involve actual parsing of argument values but only checking if an id can be found. Is this such a strong design decision, whether we call a static function of format_parse or use std::visit?

But the design doesn't add up in my opinion. If I understand it correctly, then right now the API requires to call parse in order to validate and set the parameters with the options from the command line or to print the help page. So the design deliberately forces us to call parse after we added the options to the parser. Everything is deferred. Accordingly, in the proper design the check if an option was set should be put after the parse method was called. The reason is that the interface should only work for valid options and throw if you try to ask for an option that wasn't added to the parser. But this can only be known after the option was added, and parse functions here as the terminator. So that's why I think that the current solution does not reflect the design of the argument parser.